File: analysis_engine/utils/dates.py
"""Date selection helpers for market data frames."""

from __future__ import annotations

from typing import Any

import pandas as pd


def row_on_or_after(frame: pd.DataFrame, value: Any) -> pd.Series:
    """Return the first row whose date is on or after the requested date."""
    rows = frame[frame["date"] >= pd.Timestamp(value)]
    if rows.empty:
        raise ValueError(f"No row on or after {pd.Timestamp(value).date()}")
    return rows.iloc[0]


def row_on_or_before(frame: pd.DataFrame, value: Any) -> pd.Series:
    """Return the last row whose date is on or before the requested date."""
    rows = frame[frame["date"] <= pd.Timestamp(value)]
    if rows.empty:
        raise ValueError(f"No row on or before {pd.Timestamp(value).date()}")
    return rows.iloc[-1]

File: lib/analysis_engine/utils/dates.js
export function row_on_or_after(rows, value) {
  const target = new Date(value).getTime();
  const row = rows.find((item) => new Date(item.date).getTime() >= target);
  if (!row) throw new Error(`No row on or after ${format_date(value)}`);
  return row;
}

export function row_on_or_before(rows, value) {
  const target = new Date(value).getTime();
  for (let index = rows.length - 1; index >= 0; index -= 1) {
    const row = rows[index];
    if (new Date(row.date).getTime() <= target) return row;
  }
  throw new Error(`No row on or before ${format_date(value)}`);
}

function format_date(value) {
  return new Date(value).toISOString().slice(0, 10);
}
